### PR TITLE
Feature/cldc 959 multiple tenant questions

### PR DESCRIPTION
**Pull request by kiddhustle**
TODO fix person_x_info string interpolation

https://github.com/communitiesuk/mhclg-data-collection-beta/blob/feature/CLDC-959-multiple-tenatnt-questions/config/forms/2021_2022.json#L1569

### CODE REVIEW

File: app/models/constants/case_log.rb [hunk 1]
```diff
@@ -77,23 +85,23 @@ module Constants::CaseLog
 
   NATIONAL = {
     "UK national resident in UK" => 1,
-    "A current or former reserve in the UK Armed Forces (excluding National Service)" => 100,
+    # "A current or former reserve in the UK Armed Forces (excluding National Service)" => 100,
```
Comment (kosiakkatrina) [inline app/models/constants/case_log.rb]: Is this meant to be commented out?

File: db/migrate/20220211155138_add_age_known_to_case_logs.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  def change
+    change_table :case_logs, bulk: true do |t|
+      t.column :age1_known, :string
+      t.column :age2_known, :string
+      t.column :age3_known, :string
+      t.column :age4_known, :string
+      t.column :age5_known, :string
+      t.column :age6_known, :string
+      t.column :age7_known, :string
+      t.column :age8_known, :string
```
Comment (kosiakkatrina) [inline db/migrate/20220211155138_add_age_known_to_case_logs.rb]: All of our other `_known` columns (like `la_known`, `postcode_known`) are integers, I wonder if we would want to keep it the same across these ones as well? 🤔

File: spec/models/case_log_spec.rb [hunk 3]
```diff
@@ -459,6 +459,7 @@
 
       it "correctly derives and saves totelder" do
         record_from_db = ActiveRecord::Base.connection.execute("select totelder from case_logs where id=#{household_case_log.id}").to_a[0]
+        puts record_from_db
```
Comment (kosiakkatrina) [inline spec/models/case_log_spec.rb]: Is this meant to be here?

File: config/forms/2021_2022.json [hunk 4]
```diff
@@ -1172,85 +1187,252 @@
                 }
               }
             },
-            "tenant_ethnic_group": {
+            "person_1_ethnic_group": {
               "header": "",
               "description": "",
               "questions": {
-                "ethnic": {
+                "ethnic_group": {
```
Comment (baarkerlounger) [inline config/forms/2021_2022.json:1194]: Do we need to change the question key? These are generally chosen specifically to match ADDs data dictionary (since the question name matches the DB field name)

File: config/forms/2021_2022.json [hunk 5]
```diff
@@ -1134,10 +1134,25 @@
               "header": "",
               "description": "",
               "questions": {
+                "age1_known": {
+                  "header": "Do you know the lead tenant’s age?",
```
Comment (kosiakkatrina) [inline config/forms/2021_2022.json:1138]: I think the first question should be `How many people live in the household for this letting?` ?

File: config/forms/2021_2022.json [hunk 6]
```diff
@@ -1134,10 +1134,25 @@
...
+                  },
+                  "conditional_for": {
+                    "age1": ["Yes"]
+                  }
+                },
                 "age1": {
-                  "check_answer_label": "Tenant’s age",
-                  "header": "What is the tenant’s age?",
-                  "hint_text": "",
+                  "header": "Do you know the lead tenant’s age?",
```
Comment (kosiakkatrina) [inline config/forms/2021_2022.json:1154]: Should this be just `Age` or something like that?

File: config/forms/2021_2022.json [hunk 7]
```diff
@@ -1975,20 +2627,38 @@
...
+            },
+            "person_8_sex": {
+              "header": "",
+              "description": "",
+              "questions": {
                 "sex8": {
                   "check_answer_label": "Person 8’s gender",
                   "header": "Which of these best describes person 8’s gender identity?",
-                  "hint_text": "",
+                  "hint_text": "The lead tenant is the person in the household who does the most paid work. If several people do the same paid work, the lead tenant is whoever is the oldest.",
```
Comment (baarkerlounger) [inline config/forms/2021_2022.json]: Is this hint text relevant to Person 8's gender?

File: config/forms/2021_2022.json [hunk 8]
```diff
@@ -2004,9 +2674,18 @@
...
+                "details_known_8": "Yes"
+              }]
+            },
+            "person_8_economic_status": {
+              "header": "",
+              "description": "",
+              "questions": {
                 "ecstat8": {
-                  "check_answer_label": "Person 8’s Work",
+                  "check_answer_label": "Person 8's Work",
```
Comment (baarkerlounger) [inline config/forms/2021_2022.json]: This probably needs reverting? I think we want `’` consistently in all output

File: config/forms/2021_2022.json [hunk 9]
```diff
@@ -1378,39 +1561,56 @@
...
-                    "relat8": ">6",
-                    "age8": ">6",
-                    "sex8": ">6",
-                    "ecstat8": ">6"
+                  "step": 1
+                }
+              }
+            },
+            "person_2_info": {
+              "header": "You’ve given us the details for 1 of #{@case_log.other_hhmemb + 1} people in the household",
```
Comment (kosiakkatrina) [inline config/forms/2021_2022.json]: This doesn't seem to work for me

File: spec/fixtures/exports/case_logs.xml [hunk 10]
```diff
@@ -158,6 +158,23 @@
     <previous_postcode_known>1</previous_postcode_known>
     <previous_la_known/>
     <is_previous_la_inferred>false</is_previous_la_inferred>
+    <age1_known/>
```
Comment (baarkerlounger) [inline spec/fixtures/exports/case_logs.xml]: I don't think we want to be exporting any of these since they're just specific to the form UI not the other entry methods.

File: config/forms/2021_2022.json (hunk 4 above)
Comment (baarkerlounger) [reply config/forms/2021_2022.json:1194]: Oh I see, sorry.

The saved answer we want to export as `ethnic` now comes as a follow on question. :heavy_check_mark:

File: spec/fixtures/exports/case_logs.xml (hunk 10 above)
Comment (kiddhustle) [reply spec/fixtures/exports/case_logs.xml]: OK I think I saw some errors from the the tests/linting but I think they might have been around diff I can remove these

File: spec/models/case_log_spec.rb (hunk 3 above)
Comment (kiddhustle) [reply spec/models/case_log_spec.rb]: nope!

File: spec/fixtures/exports/case_logs.xml (hunk 10 above)
Comment (kiddhustle) [reply spec/fixtures/exports/case_logs.xml]: @baarkerlounger Is it all of the new fields that need omitting of just the **_known** fields

Comment (baarkerlounger) [reply spec/fixtures/exports/case_logs.xml]: I think just the `_known` fields.

File: spec/fixtures/exports/case_logs.xml [hunk 11]
```diff
@@ -158,6 +158,23 @@
...
+    <age2_known/>
+    <age3_known/>
+    <age4_known/>
+    <age5_known/>
+    <age6_known/>
+    <age7_known/>
+    <age8_known/>
+    <ethnic_group/>
```
Comment (baarkerlounger) [inline spec/fixtures/exports/case_logs.xml]: Should this be `ethnic` rather than `ethnic_group`? That's the field that stores the final information ADD want right?

File: app/models/constants/case_log.rb (hunk 1 above)
Comment (kiddhustle) [reply app/models/constants/case_log.rb]: we don't have it as an option any more. now removed

Comment (baarkerlounger) [reply app/models/constants/case_log.rb]: Just to confirm is that change intended to be for this years form or next years form? (@paulrobertlloyd ) because changing that this year surely changes data continuity?

File: db/migrate/20220211155138_add_age_known_to_case_logs.rb (hunk 2 above)
Comment (kiddhustle) [reply db/migrate/20220211155138_add_age_known_to_case_logs.rb]: amended

File: app/models/constants/case_log.rb (hunk 1 above)
Comment (paulrobertlloyd) [reply app/models/constants/case_log.rb]: As far as I can tell, this option shouldn’t be in this list, for either collection year.

File: spec/fixtures/exports/case_logs.xml (hunk 10 above)
Comment (kiddhustle) [reply spec/fixtures/exports/case_logs.xml]: Do I need to amend  the export service to omit these fields then?

**app/services/exports/case_log_export_service.rb**

File: app/models/constants/case_log.rb (hunk 1 above)
Comment (paulrobertlloyd) [reply app/models/constants/case_log.rb]: These are the nationality options from the 2021/22 form:

Screenshot 2022-02-21 at 11 01 19

File: spec/models/case_log_spec.rb (hunk 3 above)
Comment (kiddhustle) [reply spec/models/case_log_spec.rb]: removed

File: config/forms/2021_2022.json (hunk 8 above)
Comment (kiddhustle) [reply config/forms/2021_2022.json]: reverted

File: config/forms/2021_2022.json (hunk 7 above)
Comment (kiddhustle) [reply config/forms/2021_2022.json]: aah no. it should just be on the lead tenant questions not the additional tenants

File: app/models/constants/case_log.rb (hunk 1 above)
Comment (baarkerlounger) [reply app/models/constants/case_log.rb]: :man_facepalming:  Totally missed that this was the nationality question :+1:

File: spec/fixtures/exports/case_logs.xml (hunk 10 above)
Comment (baarkerlounger) [reply spec/fixtures/exports/case_logs.xml]: Yes I think so

File: config/forms/2021_2022.json (hunk 7 above)
Comment (kiddhustle) [reply config/forms/2021_2022.json]: amended

File: config/forms/2021_2022.json (hunk 5 above)
Comment (kiddhustle) [reply config/forms/2021_2022.json:1138]: You're right it should

File: db/migrate/20220211155138_add_age_known_to_case_logs.rb (hunk 2 above)
Comment (kiddhustle) [reply db/migrate/20220211155138_add_age_known_to_case_logs.rb]: @kosiakkatrina @baarkerlounger I changed these to integers but now the depends_on conditional page doesn't work. I set the depends_on value to integers to match (0 = No, 1 = Yes)

Do we have any examples of it working for integers values?

File: config/forms/2021_2022.json (hunk 7 above)
Comment (kiddhustle) [reply config/forms/2021_2022.json]: removed

File: config/forms/2021_2022.json (hunk 6 above)
Comment (kiddhustle) [reply config/forms/2021_2022.json:1154]: For the lead tenant the text has changed:

https://miro.com/app/board/o9J_lqoxxRo=/

File: spec/fixtures/exports/case_logs.xml (hunk 11 above)
Comment (kiddhustle) [reply spec/fixtures/exports/case_logs.xml]: this have been split across two fields now but it could be omitted if necessary

Comment (kiddhustle) [reply spec/fixtures/exports/case_logs.xml]: now omitted

Comment (baarkerlounger) [reply spec/fixtures/exports/case_logs.xml]: Right. Those two field will need to be recombined into the original field then (and that exported) so the data sent to CDS remains the same.

File: spec/fixtures/exports/case_logs.xml (hunk 10 above)
Comment (kiddhustle) [reply spec/fixtures/exports/case_logs.xml]: amended

File: config/forms/2021_2022.json (hunk 9 above)
Comment (kiddhustle) [reply config/forms/2021_2022.json]: content  changed

File: spec/fixtures/exports/case_logs.xml (hunk 11 above)
Comment (kiddhustle) [reply spec/fixtures/exports/case_logs.xml]: amended

File: app/models/constants/case_log.rb [hunk 12]
```diff
@@ -0,0 +1,1134 @@
+module Constants::CaseLog
```
Comment (dushan-madetech) [inline app/models/constants/case_log.rb]: I think this file was removed in main so it shouldn't be in this PR? 🤔

File: Gemfile.lock [hunk 13]
```diff
@@ -23,10 +23,10 @@ GIT
 
 GIT
   remote: https://github.com/tagliala/activeadmin.git
-  revision: 000dd9425860fbadc72d4e08b893be098f5dff9b
+  revision: 871c6af8dec41f764a9f08bf04b5422971a23e32
```
Comment (baarkerlounger) [inline Gemfile.lock]: Think this disappears if you rebase?

File: app/services/exports/case_log_export_service.rb [hunk 14]
```diff
@@ -45,13 +45,27 @@ def retrieve_case_logs
...
-          value = case_log.read_attribute_before_type_cast(key)
-          form << doc.create_element(key, value)
+          # omit _known fileds from export
+          if key.starts_with?("details_known_") || pattern_age.match(key) || omitted_attrs.include?(key)
+            next
+          else
+            # combine ethnic fields
+            # rubocop:disable Style/ConditionalAssignment
+            if key == "ethnic"
+              value = case_log.read_attribute_before_type_cast("ethnic") || case_log.read_attribute_before_type_cast("ethnic_group")
```
Comment (baarkerlounger) [inline app/services/exports/case_log_export_service.rb]: I think we should be combining these in the model (like we do for other derived variables), not just in the export?

File: app/models/constants/case_log.rb (hunk 12 above)
Comment (kiddhustle) [reply app/models/constants/case_log.rb]: removed

File: app/services/exports/case_log_export_service.rb (hunk 14 above)
Comment (kiddhustle) [reply app/services/exports/case_log_export_service.rb]: sorry only just saw this. So via a `get_` method?

File: app/models/constants/case_log.rb (hunk 12 above)
Comment (baarkerlounger) [reply app/models/constants/case_log.rb]: Looks like it's still here? (This whole file)

File: app/services/exports/case_log_export_service.rb (hunk 14 above)
Comment (baarkerlounger) [reply app/services/exports/case_log_export_service.rb]: Something like:

`self.ethnic = ethnic || ethnic_group`

In the before validation. (Assuming `ethnic` is the fieldname that should be exported).

Comment (kiddhustle) [reply app/services/exports/case_log_export_service.rb]: oh OK so in this method:

`def set_derived_fields!`

Comment (baarkerlounger) [reply app/services/exports/case_log_export_service.rb]: Yep

File: app/services/exports/case_log_export_service.rb [hunk 15]
```diff
@@ -45,13 +45,20 @@ def retrieve_case_logs
...
+      pattern_age = /age\d_known/
+      omitted_attrs = %w[ethnic_group]
 
       case_logs.each do |case_log|
         form = doc.create_element("form")
         doc.at("forms") << form
         case_log.attributes.each do |key, _|
-          value = case_log.read_attribute_before_type_cast(key)
-          form << doc.create_element(key, value)
+          # omit _known fileds from export
```
Comment (baarkerlounger) [inline app/services/exports/case_log_export_service.rb]: Spelling on fields. Could also possibly wrap that in a method to avoid the need for a comment? Something like:

```
if omitted_field?(key)
```

**Review by baarkerlounger — APPROVED** [review]

**Comment by kiddhustle** [comment]
> Convention in Ruby is generally to have method name end in `?` if it returns a boolean

I get this lint error when I do this.

Is this down to my IDE needing configuration?

```
app/services/exports/case_log_export_service.rb:39:37: W: Lint/Syntax: invalid character syntax; use ?\n
(Using Ruby 3.1 parser; configure using TargetRubyVersion parameter, under AllCops)
    def is_omitted_field(field_name)? ...
                                    ^
app/services/exports/case_log_export_service.rb:39:37: E: Lint/Syntax: unexpected token tEH
(Using Ruby 3.1 parser; configure using TargetRubyVersion parameter, under AllCops)
    def is_omitted_field(field_name)?
```

@baarkerlounger I realised the `?` should be before the braces

**Review by baarkerlounger — APPROVED** [review]
